lib/spack/spack/test/stage.py

```diff
@@ -733,82 +673,87 @@ def test_resolve_paths(self):
+        temp_has_user = user in can_tmpdir.split(os.sep)
         paths = [os.path.join(tmp, 'stage'), os.path.join(tmp, '$user')]
-        can_paths = [canonicalize_path(paths[0]), canonicalize_path(tmp)]
-        user = getpass.getuser()
-        if user not in can_paths[1].split(os.path.sep):
-            can_paths[1] = os.path.join(can_paths[1], user)
+        can_paths = [canonicalize_path(p) for p in paths]
+
+        if temp_has_user:
+            can_paths[1] = can_tmpdir
+        else:
+            can_paths[0] = os.path.join(can_paths[0], user)
+
         assert spack.stage._resolve_paths(paths) == can_paths
 
-    def test_get_stage_root_bad_path(self, clear_stage_root, bad_stage_path):
+    def test_get_stage_root_bad_path(self, clear_stage_root):
         """Ensure an invalid stage path root raises a StageError."""
-        with pytest.raises(spack.stage.StageError,
-                           match="No accessible stage paths in"):
-            assert spack.stage.get_stage_root() is None
+        with spack.config.override('config:build_stage', '/no/such/path'):
+            with pytest.raises(spack.stage.StageError,
+                               match="No accessible stage paths in"):
+                spack.stage.get_stage_root()
 
         # Make sure the cached stage path values are unchanged.
         assert spack.stage._stage_root is None
 
-    def test_get_stage_root_non_user_path(self, clear_stage_root,
-                                          non_user_path_for_stage):
-        """Ensure a non-user stage root includes the username."""
-        # The challenge here is whether the user has access to the standard
-        # non-user path.  If not, the path should still appear in the error.
-        try:
-            path = spack.stage.get_stage_root()
-            assert getpass.getuser() in path.split(os.path.sep)
+    @pytest.mark.parametrize(
+        'path,purged', [('stage-1234567890abcdef1234567890abcdef', True),
+                        ('stage-abcdef12345678900987654321fedcba', True),
+                        ('stage-a1b2c3', False)])
+    def test_stage_purge(self, tmpdir, clear_stage_root, path, purged):
+        """Test purging of stage directories."""
+        stage_dir = tmpdir.join('stage')
+        stage_path = str(stage_dir)
 
-            # Make sure the cached stage path values are changed appropriately.
-            assert spack.stage._stage_root == path
-        except OSError as e:
-            expected = os.path.join(_non_user_root, getpass.getuser())
-            assert expected in str(e)
+        test_dir = stage_dir.join(path)
+        test_dir.ensure(dir=True)
+        test_path = str(test_dir)
 
-            # Make sure the cached stage path values are unchanged.
-            assert spack.stage._stage_root is None
+        with spack.config.override('config:build_stage', stage_path):
+            stage_root = spack.stage.get_stage_root()
+            assert stage_path == stage_root
```

Comment:
I think `test_get_stage_root_in_spack` is already testing this (i.e. that the override is working properly) so I don't think that needs to be checked in this test.

Reply:
What's the harm in confirming it is as expected before performing `purge`?

Reply:
Each test should focus on testing one thing (or as little as possible). Someone reading this test could get confused about how the assert confirms the behavior the test is focused on.

Reply:
Do you have a reference for this statement?

Testing is very dependent on the structure/modularity of the code being tested.

Test checks depend on the nature (and granularity) of the test.  For example, unit tests of functions/routines/methods should be limited to the associated side-effect(s), for example.  (Ideally you'd have 'all' "happy path" and "unhappy" path tests for every function/routine/method covered.)

Tests that build up functionality may need more checks.

In this particular case, it combined `get_stage_root` and the stage `purge` functions.  I am not comfortably with the idea of not ensuring that `get_stage_root` is returning the expected path before calling `purge` not because of the existing functionality but as "protection" against future changes.